### PR TITLE
Feat/nlp sample filter by entities

### PR DESCRIPTION
**Pull request by marrouchi**
# Motivation

This PR introduces first-class support for retrieving NlpSample documents that contain a specific set of entity and/or value patterns (e.g. “find all samples tagged with entity `intent` and value `greeting`).

# Type of change:

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by yassinedorbozgithub — APPROVED** [review]
LGTM :rocket: 

image

**Review by yassinedorbozgithub — COMMENTED** [review]
LGTM :rocket: 
image

**Comment by MohamedAliBouhaouala** [comment]
Screenshot from 2025-06-09 17-24-47

Freezes indefinitely when no keywords are provided. Also performance tends to suffer when using a very large dump

**Comment by marrouchi** [comment]
> Freezes indefinitely when no keywords are provided. Also performance tends to suffer when using a very large dump

We have already a ticket to address slow queries by adding indexes. I would suggest it would be handled in a seperate PR.

File: api/src/nlp/services/nlp-sample.service.ts [hunk 1]
```diff
@@ -47,11 +50,126 @@ export class NlpSampleService extends BaseService<
...
+      page,
+      projection,
+    );
+  }
+
+  /**
+   * Same as `findByPatterns`, but also populates all relations declared
+   * in the repository (`populatePaths`).
+   *
+   * @param criteras      `{ filters, patterns }`
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.ts]: ``criteria`` instead of ``criteras``

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 2]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
       expect(extractSpy).not.toHaveBeenCalled();
     });
   });
+
+  describe('findByPatterns', () => {
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:369]: add a test case when `patterns` is empty

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 3]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+        { filters: {}, patterns },
+        page,
+      );
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result.length).toBe(1);
+    });
+  });
+
+  describe('findByPatternsAndPopulate', () => {
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:435]: Same here. Add a test case when `patterns` is empty

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 4]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+        { filters: {}, patterns },
+        page,
+      );
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result.length).toBe(1);
+    });
+  });
+
+  describe('countByPatterns', () => {
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:503]: Same here

File: api/src/utils/generics/base-repository.ts [hunk 5]
```diff
@@ -283,6 +362,18 @@ export abstract class BaseRepository<
...
+   * - Returns `null` immediately when `criteria` is falsy.
+   * - Optional `options` are passed to `plainToClass`.
+   * - Optional `projection` limits returned fields.
+   *
+   * @param criteria   Document `_id` **or** Mongo filter.
+   * @param options    Class-transform options.
+   * @param projection Optional Mongo projection.
+   * @returns Promise resolving to the found entity or `null`.
+   */
   async findOne(
```
Comment (MohamedAliBouhaouala) [inline api/src/utils/generics/base-repository.ts:377]: No return type is needed here like applied in `findOneAndPopulate` method ?

File: api/src/utils/generics/base-repository.ts [hunk 6]
```diff
@@ -544,6 +758,18 @@ export abstract class BaseRepository<
...
+   * Update **many** documents at once.
+   *
+   * - Applies `$set` with the supplied `dto`.
+   * - When `options.shouldFlatten` is true, flattens the payload first.
+   * - Does **not** run the validation / event hooks (use `updateOne` for that).
+   *
+   * @param filter  Mongo filter selecting the documents to update.
+   * @param dto     Update payload.
+   * @param options `{ shouldFlatten?: boolean }`.
+   * @returns MongoDB `UpdateWriteOpResult` describing the operation outcome.
```
Comment (MohamedAliBouhaouala) [inline api/src/utils/generics/base-repository.ts]: @return Promise that resolves etc etc

File: api/src/nlp/repositories/nlp-sample.repository.ts [hunk 7]
```diff
@@ -32,11 +45,243 @@ export class NlpSampleRepository extends BaseRepository<
...
+      value: new Types.ObjectId(id),
+    }));
+
+    return [
+      {
+        $match: {
+          // @todo: think of a better way to handle language to objectId conversion
+          // This is a workaround for the fact that language is stored as an ObjectId
+          // in the database, but we want to filter by its string representation.
+          ...filters,
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/repositories/nlp-sample.repository.ts]: can we open a separate issue for this one just in case?

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts [hunk 8]
```diff
@@ -181,6 +182,40 @@ describe('NlpSampleController', () => {
...
+        ['language', 'entities'],
+        {},
+        patterns,
+      );
+      // Should only return samples matching the pattern
+      const nlpSamples = await nlpSampleService.findByPatternsAndPopulate(
+        { filters: {}, patterns },
+        pageQuery,
+      );
+      expect(result).toEqualPayload(nlpSamples);
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/controllers/nlp-sample.controller.spec.ts:202]: Add a check to see which function was called
```ts     
expect(nlpSampleService.findByPatternsAndPopulate).toHaveBeenCalled();
```

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts [hunk 9]
```diff
@@ -181,6 +182,40 @@ describe('NlpSampleController', () => {
...
+      ];
+      const result = await nlpSampleController.findPage(
+        pageQuery,
+        ['language', 'entities'],
+        {},
+        patterns,
+      );
+      expect(Array.isArray(result)).toBe(true);
+      expect(result).toHaveLength(0);
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/controllers/nlp-sample.controller.spec.ts:222]: same case here

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts [hunk 10]
```diff
@@ -439,4 +474,30 @@ describe('NlpSampleController', () => {
...
     });
   });
+
+  describe('filterCount', () => {
+    it('should count the nlp samples without patterns', async () => {
+      const filters = { text: 'Hello' };
+      const result = await nlpSampleController.filterCount(filters, []);
+      expect(result).toEqual({ count: 1 });
+    });
+
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/controllers/nlp-sample.controller.spec.ts:492]: Add a check to see which function was called 
```ts
expect(nlpSampleService.countByPatterns).toHaveBeenCalled();
```

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts [hunk 11]
```diff
@@ -439,4 +474,30 @@ describe('NlpSampleController', () => {
...
+    it('should count the nlp samples with patterns', async () => {
+      const filters = { text: 'Hello' };
+      const patterns: NlpValueMatchPattern[] = [
+        { entity: 'intent', match: 'value', value: 'greeting' },
+      ];
+      const result = await nlpSampleController.filterCount(filters, patterns);
+      expect(result).toEqual({ count: 1 });
+    });
+
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/controllers/nlp-sample.controller.spec.ts:503]: same here

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts [hunk 12]
```diff
@@ -439,4 +474,30 @@ describe('NlpSampleController', () => {
...
+    it('should return zero count when no samples match the filters and patterns', async () => {
+      const filters = { text: 'Nonexistent' };
+      const patterns: NlpValueMatchPattern[] = [
+        { entity: 'intent', match: 'value', value: 'nonexistent' },
+      ];
+      const result = await nlpSampleController.filterCount(filters, patterns);
+      expect(result).toEqual({ count: 0 });
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/controllers/nlp-sample.controller.spec.ts:511]: same here

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 13]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+      ];
+
+      const result = await nlpSampleService.findByPatterns(
+        { filters: {}, patterns },
+        undefined,
+      );
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result[0].text).toBe('Hello');
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:395]: perform check to see which functions were called 
```ts
expect(nlpSampleRepository.findByEntities).toHaveBeenCalled();
expect(nlpValueService.findByPatterns).toHaveBeenCalled();

```
might as well mock ``NlpValueService``

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 14]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+      ];
+
+      const result = await nlpSampleService.findByPatterns(
+        { filters: {}, patterns },
+        undefined,
+      );
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result).toHaveLength(0);
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:413]:
```ts
expect(nlpSampleRepository.findByEntities).toHaveBeenCalledTimes(0);
expect(nlpValueService.findByPatterns).toHaveBeenCalled();
```

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 15]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result.length).toBeGreaterThan(0);
+      result.forEach((sample) => {
+        expect(sample).toBeInstanceOf(NlpSampleFull);
+        expect(sample.entities).toBeDefined();
+        expect(Array.isArray(sample.entities)).toBe(true);
+        expect(sample.language).toBeDefined();
+      });
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:454]: perform method calls checks
```ts 
expect(nlpValueService.findByPatterns).toHaveBeenCalled();
expect(nlpSampleRepository.findByEntitiesAndPopulate).toHaveBeenCalled();
```

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 16]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+      ];
+
+      const result = await nlpSampleService.findByPatternsAndPopulate(
+        { filters: {}, patterns },
+        undefined,
+      );
+
+      expect(Array.isArray(result)).toBe(true);
+      expect(result).toHaveLength(0);
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:481]:
```ts
expect(nlpValueService.findByPatterns).toHaveBeenCalled();
expect(nlpSampleRepository.findByEntitiesAndPopulate).toHaveBeenCalledTimes(0);
```

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 17]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+      ];
+
+      const count = await nlpSampleService.countByPatterns({
+        filters: {},
+        patterns,
+      });
+
+      expect(typeof count).toBe('number');
+      expect(count).toBe(2);
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:520]: perform methods calls check 
```ts
expect(nlpValueService.findByPatterns).toHaveBeenCalled();
expect(nlpSampleRepository.countByEntities).toHaveBeenCalled();
```

File: api/src/nlp/services/nlp-sample.service.spec.ts [hunk 18]
```diff
@@ -360,4 +362,153 @@ describe('NlpSampleService', () => {
...
+        { entity: 'intent', match: 'value', value: 'nonexistent' },
+      ];
+
+      const count = await nlpSampleService.countByPatterns({
+        filters: {},
+        patterns,
+      });
+
+      expect(count).toBe(0);
+    });
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/services/nlp-sample.service.spec.ts:547]:
```ts
expect(nlpValueService.findByPatterns).toHaveBeenCalled();
expect(nlpSampleRepository.countByEntities).toHaveBeenCalledTimes(0);
```

**Review by MohamedAliBouhaouala — CHANGES_REQUESTED** [review]

File: api/src/nlp/repositories/nlp-sample.repository.ts [hunk 19]
```diff
@@ -32,11 +45,243 @@ export class NlpSampleRepository extends BaseRepository<
...
+              },
+            ],
+          },
+        },
+      },
+
+      //drop helper array if you don’t need it downstream
+      { $project: { entities: 0, sampleentities: 0 } },
+    ];
+  }
```
Comment (MohamedAliBouhaouala) [inline api/src/nlp/repositories/nlp-sample.repository.ts]: We should refactor and retailor this method. Too long almost unreadable for the intiated. Example: 
```ts
 buildFindByEntitiesStages(
    {
      filters,
      values,
    }: {
      filters: TFilterQuery<NlpSample>;
      values: NlpValue[];
    },
    requireAll = true,
  ): PipelineStage[] {
    const requiredPairs = values.map(({ id, entity }) => ({
      entity: new Types.ObjectId(entity),
      value: new Types.ObjectId(id),
    }));

    // Handle language filter if present
    const processedFilters = this.convertLanguageFilters(filters);
    return [
      // Initial filter
      { $match: processedFilters },

      // Lookup matching entities
      {
        $lookup: {
          from: 'nlpsampleentities',
          localField: '_id',
          foreignField: 'sample',
          as: 'matches',
          pipeline: [{ $match: { $or: requiredPairs } }],
        },
      },

      // Filter by match count
      {
        $match: {
          $expr: requireAll
            ? { $eq: [{ $size: '$matches' }, requiredPairs.length] }
            : { $gt: [{ $size: '$matches' }, 0] },
        },
      },

      // Clean up
      { $project: { matches: 0 } },
    ];
  }
 private convertLanguageFilters(filter): TFilterQuery<NlpSample> {
    if (!filter) return filter;

    // Handle direct language filter
    if (filter.language && typeof filter.language === 'string') {
      return {
        ...filter,
        language: new Types.ObjectId(filter.language),
      };
    }

    // Handle $and conditions
    if (Array.isArray(filter.$and)) {
      return {
        ...filter,
        $and: filter.$and.map((f) => this.convertLanguageFilters(f)),
      };
    }

    // Handle $or conditions
    if (Array.isArray(filter.$or)) {
      return {
        ...filter,
        $or: filter.$or.map((f) => this.convertLanguageFilters(f)),
      };
    }

    return filter;
  }
```

File: api/src/nlp/repositories/nlp-sample.repository.ts (hunk 7 above)
Comment (marrouchi) [reply api/src/nlp/repositories/nlp-sample.repository.ts]: yes @MohamedAliBouhaouala Can you please open a ticket about this ?

File: api/src/nlp/controllers/nlp-sample.controller.spec.ts (hunk 8 above)
Comment (marrouchi) [reply api/src/nlp/controllers/nlp-sample.controller.spec.ts:202]: It's not relevant I think, we are calling it already in the previous line. Maybe you mean repository ?

Comment (MohamedAliBouhaouala) [reply api/src/nlp/controllers/nlp-sample.controller.spec.ts:202]: It's either ``findByPatternsAndPopulate`` or ``findByPatterns``. That is why I proposed a test for method call in this case.
I also think the repository call should be tested at the service layer as previously suggested.
